File: archive/flutter-app/lib/core/office_media/image_converter.dart
import 'dart:typed_data';

import 'package:image/image.dart' as img;

import '../ports/i_tool_use_case.dart';

/// Target raster format for [ImageConverter].
enum ImageOutputFormat {
  jpeg('JPEG', 'jpg'),
  png('PNG', 'png'),
  webp('WebP', 'webp');

  const ImageOutputFormat(this.label, this.extension);

  final String label;
  final String extension;
}

class ImageConversionInput {
  const ImageConversionInput({
    required this.sourceBytes,
    required this.targetFormat,
    this.quality = 85,
    this.maxWidth,
    this.maxHeight,
  });

  /// Raw bytes of the source image, in any format `package:image` can
  /// auto-detect (JPEG, PNG, GIF, BMP, TGA, WebP-lossless, ...).
  final Uint8List sourceBytes;

  final ImageOutputFormat targetFormat;

  /// 0-100. Used verbatim for JPEG (libjpeg-style quality). For PNG this is
  /// remapped to the 0-9 zlib compression-*effort* level (PNG is always
  /// lossless — a higher "quality" here just means more CPU spent finding a
  /// smaller file, not lower fidelity). Ignored for WebP: see the doc
  /// comment on [ImageConverter] for why.
  final int quality;

  /// Optional resize bounds, applied before encoding. Aspect ratio is
  /// preserved (matching `package:image`'s `copyResize(maintainAspect:
  /// true)`), so the output may be smaller than the exact box requested.
  final int? maxWidth;
  final int? maxHeight;
}

class ImageConversionResult {
  const ImageConversionResult({
    required this.outputBytes,
    required this.outputFormat,
    required this.sourceByteSize,
    required this.outputByteSize,
    required this.sourceWidth,
    required this.sourceHeight,
    required this.outputWidth,
    required this.outputHeight,
  });

  final Uint8List outputBytes;
  final ImageOutputFormat outputFormat;

  final int sourceByteSize;
  final int outputByteSize;

  final int sourceWidth;
  final int sourceHeight;
  final int outputWidth;
  final int outputHeight;

  /// outputByteSize / sourceByteSize. Less than 1.0 means the output is
  /// smaller than the source; e.g. 0.4 means the output is 40% of the
  /// original size (a 60% reduction).
  double get compressionRatio =>
      sourceByteSize == 0 ? 0 : outputByteSize / sourceByteSize;

  int get bytesSaved => sourceByteSize - outputByteSize;

  /// Positive when the output shrank, negative when it grew (can happen —
  /// e.g. re-encoding a highly-optimized small PNG as JPEG at high quality).
  double get percentSaved =>
      sourceByteSize == 0 ? 0 : (bytesSaved / sourceByteSize) * 100;
}

/// Image Format Converter & Compressor (spec 3.2): converts between JPEG,
/// PNG and WebP and optionally resizes/recompresses along the way, built
/// entirely on `package:image`'s pure-Dart codecs so it has zero platform
/// dependency and is safely unit-testable.
///
/// Source format is auto-detected from the bytes via `img.decodeImage()` —
/// callers never need to say what the input is, only what they want out.
///
/// WebP caveat: `package:image` 4.9.2's `WebPEncoder` only implements the
/// **lossless VP8L** bitstream — there is no lossy VP8 encoder and therefore
/// no quality knob for WebP output in this package. Converting *to* WebP
/// here always produces a lossless file regardless of [ImageConversionInput.
/// quality]; that field is silently ignored for that target. (Decoding WebP,
/// including lossy VP8 files produced by other tools, is still supported —
/// this limitation is encode-only.) If true lossy WebP output is ever
/// required, a native binding (e.g. `libwebp`) would be needed instead.
class ImageConverter
    implements IToolUseCase<ImageConversionInput, ImageConversionResult> {
  const ImageConverter();

  @override
  ImageConversionResult execute(ImageConversionInput input) {
    if (input.quality < 0 || input.quality > 100) {
      throw ArgumentError.value(
        input.quality,
        'quality',
        'Must be between 0 and 100',
      );
    }
    if ((input.maxWidth != null && input.maxWidth! <= 0) ||
        (input.maxHeight != null && input.maxHeight! <= 0)) {
      throw ArgumentError('maxWidth/maxHeight must be positive when given');
    }

    final decoded = img.decodeImage(input.sourceBytes);
    if (decoded == null) {
      throw FormatException(
        'Could not decode image: unrecognized or corrupt image data',
      );
    }

    var working = decoded;
    if (input.maxWidth != null || input.maxHeight != null) {
      working = img.copyResize(
        working,
        width: input.maxWidth,
        height: input.maxHeight,
        maintainAspect: true,
        interpolation: img.Interpolation.average,
      );
    }

    final Uint8List outputBytes;
    switch (input.targetFormat) {
      case ImageOutputFormat.jpeg:
        outputBytes = img.encodeJpg(working, quality: input.quality.clamp(1, 100).toInt());
      case ImageOutputFormat.png:
        outputBytes = img.encodePng(working, level: _pngLevelFromQuality(input.quality));
      case ImageOutputFormat.webp:
        // No quality parameter exists on this encoder — see class doc.
        outputBytes = img.encodeWebP(working);
    }

    return ImageConversionResult(
      outputBytes: outputBytes,
      outputFormat: input.targetFormat,
      sourceByteSize: input.sourceBytes.length,
      outputByteSize: outputBytes.length,
      sourceWidth: decoded.width,
      sourceHeight: decoded.height,
      outputWidth: working.width,
      outputHeight: working.height,
    );
  }

  /// Maps the familiar 0-100 "quality" scale onto PNG's 0-9 deflate
  /// compression-effort level, so the UI can offer a single slider even
  /// though PNG has no lossy quality concept.
  int _pngLevelFromQuality(int quality) =>
      (quality.clamp(0, 100) / 100 * 9).round().clamp(0, 9);
}

File: app/src/core/office_media/imageConverter.ts
import type { IToolUseCase } from '../ports/IToolUseCase'

/** Target raster format for {@link ImageConverter}. */
export type ImageOutputFormat = 'jpeg' | 'png' | 'webp'

export const IMAGE_OUTPUT_FORMATS: ImageOutputFormat[] = ['jpeg', 'png', 'webp']

interface FormatInfo {
  label: string
  extension: string
  mimeType: string
}

const FORMAT_INFO: Record<ImageOutputFormat, FormatInfo> = {
  jpeg: { label: 'JPEG', extension: 'jpg', mimeType: 'image/jpeg' },
  png: { label: 'PNG', extension: 'png', mimeType: 'image/png' },
  webp: { label: 'WebP', extension: 'webp', mimeType: 'image/webp' },
}

export function labelForFormat(format: ImageOutputFormat): string {
  return FORMAT_INFO[format].label
}

export function extensionForFormat(format: ImageOutputFormat): string {
  return FORMAT_INFO[format].extension
}

export function mimeTypeForFormat(format: ImageOutputFormat): string {
  return FORMAT_INFO[format].mimeType
}

export interface ImageConversionInput {
  /** Raw bytes of the source image, in any format the browser's image decoder can handle
   * (JPEG, PNG, GIF, BMP, WebP, ...). */
  sourceBytes: Uint8Array
  targetFormat: ImageOutputFormat
  /** 0-100. Mapped to Canvas's `toBlob` quality argument (0-1) for JPEG/WebP output.
   * Ignored for PNG — PNG is always lossless, and unlike `package:image`'s zlib
   * compression-*effort* level (0-9), the browser's Canvas `toBlob` has no PNG
   * compression-effort knob at all, so PNG output size is whatever the browser's
   * built-in PNG encoder produces regardless of this value. Defaults to 85. */
  quality?: number
  /** Optional resize bounds, applied before encoding. Aspect ratio is always preserved
   * (the output fits within the box, matching `package:image`'s
   * `copyResize(maintainAspect: true)`), so the output may be smaller than the exact
   * box requested. */
  maxWidth?: number
  maxHeight?: number
}

export interface ImageConversionResult {
  outputBytes: Uint8Array
  outputFormat: ImageOutputFormat

  sourceByteSize: number
  outputByteSize: number

  sourceWidth: number
  sourceHeight: number
  outputWidth: number
  outputHeight: number
}

/** outputByteSize / sourceByteSize. Less than 1.0 means the output is smaller than the
 * source; e.g. 0.4 means the output is 40% of the original size (a 60% reduction). */
export function compressionRatio(result: ImageConversionResult): number {
  return result.sourceByteSize === 0 ? 0 : result.outputByteSize / result.sourceByteSize
}

export function bytesSaved(result: ImageConversionResult): number {
  return result.sourceByteSize - result.outputByteSize
}

/** Positive when the output shrank, negative when it grew (can happen — e.g.
 * re-encoding a highly-optimized small PNG as JPEG at high quality). */
export function percentSaved(result: ImageConversionResult): number {
  return result.sourceByteSize === 0 ? 0 : (bytesSaved(result) / result.sourceByteSize) * 100
}

/** Throws if `quality` is outside the valid 0-100 range. */
export function validateQuality(quality: number): void {
  if (quality < 0 || quality > 100) {
    throw new Error(`quality must be between 0 and 100, got ${quality}`)
  }
}

/** Throws if either resize bound was supplied but isn't a positive number. */
export function validateResizeBounds(maxWidth?: number, maxHeight?: number): void {
  if ((maxWidth != null && maxWidth <= 0) || (maxHeight != null && maxHeight <= 0)) {
    throw new Error('maxWidth/maxHeight must be positive when given')
  }
}

/**
 * Computes the output dimensions for an aspect-ratio-preserving resize into an
 * optional `maxWidth`/`maxHeight` box — the same "fit within the box, keep aspect"
 * behavior as `package:image`'s `copyResize(maintainAspect: true)`. Pure and
 * synchronous so it's directly unit-testable, independent of the Canvas-dependent
 * encode step in {@link ImageConverter}.
 */
export function computeResizedDimensions(
  sourceWidth: number,
  sourceHeight: number,
  maxWidth?: number,
  maxHeight?: number,
): { width: number; height: number } {
  if (maxWidth == null && maxHeight == null) {
    return { width: sourceWidth, height: sourceHeight }
  }
  let scale = Number.POSITIVE_INFINITY
  if (maxWidth != null) scale = Math.min(scale, maxWidth / sourceWidth)
  if (maxHeight != null) scale = Math.min(scale, maxHeight / sourceHeight)
  return {
    width: Math.max(1, Math.round(sourceWidth * scale)),
    height: Math.max(1, Math.round(sourceHeight * scale)),
  }
}

/**
 * Image Format Converter & Compressor (spec 3.2): converts between JPEG, PNG and WebP
 * and optionally resizes/recompresses along the way, built on the browser's native
 * Canvas API — `createImageBitmap` to decode, `<canvas>` + `drawImage` to resize, and
 * `canvas.toBlob` to re-encode. Source format is auto-detected by the browser's
 * decoder; callers never need to say what the input is, only what they want out —
 * the same contract the Dart version's `package:image`-backed `ImageConverter` had.
 *
 * Unlike the Dart version (whose `package:image` 4.9.2 `WebPEncoder` only supports
 * lossless VP8L output, so `quality` was silently ignored for WebP targets), the
 * browser's `canvas.toBlob('image/webp', quality)` supports real lossy WebP encoding
 * — so here `quality` *does* apply to WebP output, not just JPEG. This is a genuine
 * capability improvement over the Dart core, not a regression.
 */
export class ImageConverter implements IToolUseCase<ImageConversionInput, ImageConversionResult> {
  async execute(input: ImageConversionInput): Promise<ImageConversionResult> {
    const quality = input.quality ?? 85
    validateQuality(quality)
    validateResizeBounds(input.maxWidth, input.maxHeight)

    let bitmap: ImageBitmap
    try {
      // `new Uint8Array(input.sourceBytes)` copies into a plain-`ArrayBuffer`-backed
      // view — `Blob`'s `BlobPart` type requires that concrete backing, which the
      // caller's `Uint8Array` isn't statically guaranteed to have.
      bitmap = await createImageBitmap(new Blob([new Uint8Array(input.sourceBytes)]))
    } catch {
      throw new Error('Could not decode image: unrecognized or corrupt image data')
    }

    const sourceWidth = bitmap.width
    const sourceHeight = bitmap.height
    const { width: outputWidth, height: outputHeight } = computeResizedDimensions(
      sourceWidth,
      sourceHeight,
      input.maxWidth,
      input.maxHeight,
    )

    const canvas = document.createElement('canvas')
    canvas.width = outputWidth
    canvas.height = outputHeight
    const ctx = canvas.getContext('2d')
    if (!ctx) {
      bitmap.close()
      throw new Error('Canvas 2D context is not available in this environment')
    }
    ctx.drawImage(bitmap, 0, 0, outputWidth, outputHeight)
    bitmap.close()

    const mimeType = mimeTypeForFormat(input.targetFormat)
    // Quality only affects JPEG/WebP — passing it for PNG is harmless (ignored by
    // the encoder) but omitting it is clearer about intent.
    const encoderQuality = input.targetFormat === 'png' ? undefined : quality / 100

    const blob = await new Promise<Blob>((resolve, reject) => {
      canvas.toBlob(
        (result) => (result ? resolve(result) : reject(new Error('Image encoding failed'))),
        mimeType,
        encoderQuality,
      )
    })

    const outputBytes = new Uint8Array(await blob.arrayBuffer())

    return {
      outputBytes,
      outputFormat: input.targetFormat,
      sourceByteSize: input.sourceBytes.byteLength,
      outputByteSize: outputBytes.byteLength,
      sourceWidth,
      sourceHeight,
      outputWidth,
      outputHeight,
    }
  }
}
